### DISCUSSION TITLE
Persistence keys: help actions manage their own state

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#1883] [Ideas]
This is a proposal for one particular solution to the problem of state persistence in Dagger.

## Problem

Sometimes, a component in the DAG needs to remember what they did earlier - usually so that they can undo it. This feature is called “persistent state” or “state” for short. For example, a component which creates Amazon S3 buckets needs persistent state to remember which buckets they created, in case the user wants to destroy them later.

One obvious use of persistent state is the “dagger down” feature (#309): the ability to “clean up” all side effects cause by a Dagger environment. But there are other uses as well.

One reason Dagger does not yet support state persistence, is that it is very difficult to do correctly. The more state Dagger is responsible for persisting, the more things can go wrong, and the more complex Dagger must become to manage possible errors. There is also the problem of customization: managing more state means storing more data, and different users have very different opinions on how to store their data. The bigger the data, the more opinions diverge on how to store it, and the more customizable Dagger must be to accomodate’s everyone’s opinion. This leads to more complexity.

The ideal solution would allow individual Dagger components to persist as much state as they need, while keeping data size and customization of data storage, and therefore complexity, to an absolute minumum.

## Solution

There is no way to implement state persistence in the Dagger core in a way that will 1) meet the requirements of every component in the DAG, 2) avoid the extra complexity of implementing reliable storage, 3) avoid the extra complexity of allowing customization of storage backend.

So we don’t. Instead, we implement the primitives necessary for each component in the DAG to either 1) implement its own storage, or 2) provide storage middleware for other components - all in Cue.

This requires 2 primitives:

1. Persistence keys: a CUE developer can request that Dagger generate *and persist* a unique key (perhaps a UUID, perhaps a human-friendly slug.. details are TBD), and inject it into the specified Cue value.

2. Storage middleware interface: the standard library should define a common storage API in an abstract Cue interface, and provide a reference implementation on the provider side (S3; local directory; rsync..) and consumer side (terraform, netlify..) to seed an ecosystem of storage middleware. This is crucial, for example, to allow wrapping of low-level package such as terraform without leaking abstraction. If I import a `multidb` package which happens to wrap a custom terraform configuration, it should be easy (and perhaps required) for me to provide a standard storage middleware component to store terraform state. It *should not* be locked in that I can only use, say, AWS S3.

### Cue API for persistence keys

This is still work in progress.

Example code:

```
appname: string @dagger(persistenceKey)
```

Example value:`appname: “happy-panda-3654”`

### Storage of persistence keys

Dagger is responsible for storing the persistence keys and mapping them to the correct environment and path. This is similar to how output values are persisted today.

The defining feature of this proposal is what Dagger is *not* responsible for: storing arbitrary data and making them available read-write to buildkit pipelines. This is now 100% the responsibility of the Cue developer (aka “userland”).

### Storage middleware interface

This is still work in progress. The storage middleware interface is a regular Cue interface. Any Dagger user can design and share their own interface. But only Dagger maintainers can define a standard and designate the “correct” patterns.

**Comment by jlongtine** [comment]
This is an interesting/difficult problem. I totally agree re the need. It's something that was coming up for me as I was thinking about `dagger down`: clearing state as the DAG gets unwound.

I like the idea of using CUE as the interface for this. Should simplify the interface and make it clearer. Also, the ability to define an arbitrary struct to be stored as state would be awesome.

**Comment by shykes** [comment]
This is still relevant in Dagger 0.2, although the design for a solution might be different now.
